Fix some include paths behavior

wfleming opened the pull request:
1. Support the `.markdown` extensions when filtering for files directly.
   
   The longer `.markdown` is actually the one endorsed by John Gruber, Markdown's creator. There [are other extensions in use, as well](http://daringfireball.net/linked/2014/01/08/markdown-extension), but I think these two are fine for nowsince they're by far the most used and are what GitHub supports as
   well.
2. Support directories in config-provided `include_paths`
   
   Without this, only Markdown files in the root directory would ever be analyzed. Which is most Markdown, certainly, but not all.

wfleming committed:
Support .markdown file extension when filtering

`.md` & `.markdown` are the most common extensions, and are the two
supported by GitHub. The longer `.markdown` is actually the one endorsed
by John Gruber, Markdown's creator.

There are others, as well[1], but I think these two are fine for now
since they're by far the most used and are what GitHub supports as
well.

wfleming committed:
Accept directories in include_paths

Without this, the engine would likely only end up parsing Markdown files
in the root directory. (Which, admittedly, is most Markdown files, but
not all).

The path-parsing change in issue emission was necesary because `mdl`
internally does some unusual string construction of file names when
given relative paths to analyze. This is the easiest way to get around
it.

jpignata commented on lib/cc/engine/markdownlint.rb:
```diff
@@ -33,7 +35,7 @@ def include_paths
 
         markdown_files = engine_config["include_paths"].select do |path|
-          path.end_with?(".md")
+          EXTENSIONS.include?(File.extname("path")) || path.end_with?("/")
```
Should this be `path`?

wfleming replied:
Sigh, yes. I knew I should have written more tests, but writing new fixtures seemed like more trouble than it was worth at the time.

wfleming replied:
Fixed.

jpignata commented on lib/cc/engine/markdownlint.rb:
```diff
@@ -5,6 +5,8 @@
 module CC
   module Engine
     class Markdownlint
+      EXTENSIONS = %w[.markdown .md].freeze
```
Thanks! I punted here and just went with `.md` since that was `mdl` did, but this is better.

jpignata commented on lib/cc/engine/markdownlint.rb:
```diff
@@ -33,7 +35,7 @@ def include_paths
 
         markdown_files = engine_config["include_paths"].select do |path|
-          path.end_with?(".md")
+          EXTENSIONS.include?(File.extname(path)) || path.end_with?("/")
```
Do you think this means `mdl` will try to analyze every file in these subdirectories? If so, do we need to pass `#{path}**/*.md` and `#{path}**/*.markdown` instead?

wfleming replied:
It means `mdl` will try and analyze every `*.md` file in the given directories. Which isn't quite right. So you're right, we probably need to walk the tree ourselves & get all the actual files.

pbrisbin replied:
FWIW: I'd be in favor of supporting `*.md` only, if that's what the underlying tool does and it simplifies the engine in a meaningful way (which it seems like it would). At least until we see actual user demand for `*.markdown` support.

wfleming replied:
I am a user and I use .markdown, personally. I should open a PR with mdl, though.

~will out.

On Feb 29, 2016, 07:59 -0500, pat brisbinnotifications@github.com, wrote:

> Inlib/cc/engine/markdownlint.rb(https://github.com/jpignata/codeclimate-markdownlint/pull/1#discussion_r54405898):
> 
> > @@ -33,7 +35,7 @@ def include_paths>return root unless engine_config.has_key?("include_paths")>>markdown_files = engine_config["include_paths"].select do |path|>-     path.end_with?(".md")>+     EXTENSIONS.include?(File.extname(path)) || path.end_with?("/")
> 
> FWIW: I'd be in favor of supporting_.mdonly if that's what the underlying tool does and it simplifies the engine in a meaningful way (which it seems like it would). At least until we see actual user demand for_.markdownsupport.
> 
> —
> Reply to this email directly orview it on GitHub(https://github.com/jpignata/codeclimate-markdownlint/pull/1/files#r54405898).

wfleming committed:
Use wfleming's fork of markdownlint

The fork adds support for the .markdown file extension.
Work at: https://github.com/wfleming/markdownlint/commit/8143d80a31b6f803bdc0f0b90ada3971546d167b

I've opened a PR against the parent repo, but that repo has been pretty
quiet for a couple months, so I think it's expedient to use this fork
until the parent repo becomes active again.

wfleming commented:
@jpignata I forked `markdownlint` & implemented built-in support for the `*.markdown` extension there. I've opened [a PR](https://github.com/mivok/markdownlint/pull/119) for that work, but since the parent repo has been quiet for a few months, I've switched this engine to use my fork in the interim.
